### ISSUE TITLE
File types supported?

### ISSUE THREAD

**Issue by kariohlsen** [MicrosoftDocs/Azure-RMSDocs#552]
Can you confirm that the RMS SDK for iOS supports .docx, .xlsx, and .pptx filetypes?  We are having issues with those, while .ptxt seems to work in our hands

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 31edfca6-d58e-4cb7-a55f-72665a605d31
* Version Independent ID: d2bf323c-6d14-4c64-df52-111c454a7e6a
* Content: [Azure Information Protection SDK 4.2 developer guidance](https://docs.microsoft.com/en-us/azure/information-protection/develop/core-concepts)
* Content Source: [Azure-RMSDocs/Develop/core-concepts.md](https://github.com/Microsoft/Azure-RMSDocs/blob/master/Azure-RMSDocs/Develop/core-concepts.md)
* Service: **information-protection**
* GitHub Login: @lleonard-msft
* Microsoft Alias: **alleonar**

**Comment by kariohlsen** [comment]
I do not see any APIs that support streams in the downloaded SDK that we got from https://www.microsoft.com/en-us/download/details.aspx?id=43674

Our Android developer has found some stream support in the Android SDK, but we do not see it in the iOS SDK.  Is there another place we should be downloading the SDK from?

**Comment by kkanakas** [comment]
Have you looked at the API reference ? https://msdn.microsoft.com/library/dn758306.aspx
This is the class you need to look into 
https://msdn.microsoft.com/en-us/library/mt210993.aspx

**Comment by kariohlsen** [comment]
We did see that class, but it looks like you need a MSUserPolicy to initialize the class.  We want to decrypt a file, and cannot figure out how to extract the applied MSUserPolicy from the file.  Can you give us some guidance here? Sorry if we are missing something obvious.

**Comment by kkanakas** [comment]
here is the sample app for iOS
 https://github.com/AzureAD/rms-sdk-ui-for-ios/tree/master/samples/RMSSampleApp/IPApp

let me know if run into problems here

**Comment by kariohlsen** [comment]
Thanks, we were already aware of this sample app and have been using it for a model.  However it is not compatible with the .docx files we are using. It only supports .txt, .ptxt and txt2 files. I have modified the source to accept .docx files (as a customProtectedFile), but we get an error when trying to open a .docx file in this app.

Note that this .docx file can be successfully opened in Word on Mac Word, Windows Word and iOS Word with the proper login, so it doesn’t seem to be the correct error message.


The key for us would be to be able to extract the policy from the file, and in the sample code we see in the handleCustomProtectedFile this code:



  // Read header information from protectedData and extract the  PL

    /*-------------------------------------------

     | PL length | PL | ContentSizeLength |

     -------------------------------------------*/

    NSData *plSizeData = [irmFile readDataOfLength:sizeof(NSUInteger)];

    NSUInteger plSize;

    [plSizeData getBytes:&plSize range:NSMakeRange(0, sizeof(NSUInteger))];



    // get the PL and protectedData

    NSData *pl = [irmFile readDataOfLength:plSize];

Guessing that this might be the structure for the RMS file, we tried feeding a protected docx file into this, and we get these errors:

2018-07-13 16:46:37.152549-0700 RMSGD[47570:8771976] ERROR: work thread: failed to parse the PL with error : Error Domain=RMSDomain Code=-1 "There was an unexpected error when connecting to the Rights Management service. Contact Microsoft for further assistance." UserInfo={NSLocalizedDescription=There was an unexpected error when connecting to the Rights Management service. Contact Microsoft for further assistance.,
..
2018-07-13 17:18:26.606104-0700 RMSSampleApp[7262:3229038] Consumption failed with error: Error Domain=RMSDomain Code=-7 "The publishing license for the protected content has not been set correctly. Contact the application support for further assistance."

Is the header described in the code comment the correct structure for the .docx file, and is code from the RMS Sample app to extract the policy correct? The entire file appears to be consumed when reading the pl component, so this doesn’t seem to be right.

Thanks for your help
Kari

**Comment by kariohlsen** [comment]
Yes we have already authenticated to the RMS Service.  Because we do not get the callback with a docx file, we don't get the policy we would need to pass here. For complete context, here is what has been tried.

1) [MSProtectedData protectedDatawithProtectedFile:file.path userId:self.userId, authentication callback:self consentCallback:self options:Default completionBlock:^..}];

The auth and consent callbacks have been implemented.  However those callbacks are only received with a ptxt file.  With a docx file we get the error "Failed in checkPreamble, preamble does not match expected value", and you indicated in 7/12/18 comment that file-api's are not supported for the SDK.

2)  [MSUserPolicy userPolicyWithSerializedPolicy:pl userId:self.userId  authenticationCallback:self consentCallback:self options:Default completionBlock:^(MSUserPolicy *userPolicy, NSError *error) 
- where pl is bits extracted from beginning of file per the code sample I posted on 7/15/18

Error is "failed to parse the PL with error : Error Domain=RMSDomain Code=-1 "There was an unexpected error when connecting to the Rights Management service."

3 )MSProtector *protector = [[MSProtector alloc] init];
    NSData *encryptedData =  [NSData dataWithContentsOfURL:newURL];
    NSData *outData = [protector decryptAlignedBlocksWithStartingBlockNumber:0 encryptedData:encryptedData isFinal:YES error:&error];

Error is "EXC_ARITHMETIC (code=EXC_I386_DIV subcode=0x0)"

We also ran the login before trying any of those approaches (by explicitly requesting a token with 
 [self.authContext acquireTokenWithResource:..) and got the same results. 

It seems that the API's just don't support our use case, but that seems crazy since this must be a common need. Thanks for looking into this.

**Comment by chandu4iOS** [comment]
@kariohlsen , I am running into the same set of errors which you have posted above when I treated docx file as custom protected file and extracted PL using the sample code. I too noticed that the entire file is being picked up while extracting the PL component from the file. So this makes me think either the protected docx file (protected on a windows machine with word application with a label) cannot be treated as a custom protected file or header structure for extracting the PL is incorrect. Did you manage to figure out the root cause and workarounds

**Comment by kariohlsen** [comment]
@chandu4iOS unfortunately I was not able to resolve this. From the comment on the other thread (https://github.com/AzureAD/rms-sdk-ui-for-ios/issues/1#issuecomment-450633478 , where you have also commented) we concluded that the MS Office RMS encryption was not compatible with the iOS MS RMS SDK (which seems to be designed only for custom encryption).  We had to use the Windows RMS SDK and handle RMS permissions and decryption on the Windows Server side.
